client/src/extension.ts

```diff
@@ -91,12 +94,93 @@ export function activate(context: ExtensionContext) {
 	}
 
 	// Create the language client and start the client.
-	let disposable = new LanguageClient('cfnLint', 'CloudFormation linter Language Server', serverOptions, clientOptions).start();
+	let languageClient = new LanguageClient('cfnLint', 'CloudFormation linter Language Server', serverOptions, clientOptions);
+	let disposable = languageClient.start();
+
+	languageClient.onReady().then(() => {
+		languageClient.onNotification('cfn/previewisavailable', (uri) => {
+			// This enables the "show preview" button in the top right corner
+			vscode.commands.executeCommand('setContext', 'isCfnPreviewAvailable', true);
+
+			// We don't want to show the preview if the user didn't request it. But, if the preview is visible, reload it
+			if (previews[uri]) {
+				vscode.commands.executeCommand('extension.sidePreview');
+			}
+		})
+		languageClient.onNotification('cfn/fileclosed', (uri) => {
+			// if the user closed the template itself
+			previews[uri].dispose();
+		})
+	})
 
 	// Push the disposable to the context's subscriptions so that the
 	// client can be deactivated on extension deactivation
 	context.subscriptions.push(disposable);
 
+	context.subscriptions.push(
+		vscode.commands.registerCommand('extension.sidePreview', loadSidePreview)
+	  );
+
+}
+
+function loadSidePreview() {
+	if (!window.activeTextEditor.document)
+		return;
+	let uri = window.activeTextEditor.document.uri;
+	//FIXME security concern - we should read this file from the extension's workspace
```

Comment:
not sure I understand the concern here

Reply:
It looked sketchy at first but upon reading the docs it looks like it's allowed: https://code.visualstudio.com/api/extension-guides/webview#loading-local-content

> By default, webviews can only access resources in the following locations:
> - Within your extension's install directory.
> - Within the user's currently active workspace.